### PR TITLE
Closes #2898 New Feature: Preload Links

### LINKED ISSUE #2898: Preload pages before the user click on it
## What are we proposing to do? 

With this feature, WP Rocket will preload a link before the user is ready to click on it. Initially we will be preloading as soon as the user hovers over a link that they are about to click. In the future we can iterate to be smarter about this. 

<p></p>

An example of a similar feature is available here: [<u>https://instant.page</u>](https://instant.page)

## Why are we doing this?

Before clicking on a link, a user hovers over the link for a fraction of a second. If we start preloading the page at this moment, that is time saved from the actual rendering of the page. This can result in a perceived performance gain and provide a better user experience. 

## User expectations and experience. How will we do this feature?

Users will enable this feature via a UI option in the “Preload” tab of WP Rocket. Everything else will be handled by WP Rocket automatically without further user input. 

<p></p>

**UI Proposal**

<p></p>

Location: “Preload” tab. **After “Preload Fonts”.** 

Section Title: **Preload Links**

Option Description: **Link preloading improves the perceived load time by downloading a page when a user hovers over the link. More info**

Option Name: **Enable link preloading**

## Technical Specs

A comparison of existing solutions is available here: [<u>https://wpspeedmatters.com/quicklink-vs-instant-page-vs-flying-pages/</u>](https://wpspeedmatters.com/quicklink-vs-instant-page-vs-flying-pages/)

<p></p>

Some feature recommendations that I think are useful: 

- Preload only HTML, not the images or other assets. 

- Do not preload if the mouse just passes over the link. If the user scrolls over a menu, that will be a lot of preloaded links. 

- Limit the number of preloads per second. This will also help with the previous situation where the user scrolls over a bunch of links. 

- Disable the feature if the browser uses data saver mode. Instant page does this. 

## Risks

- Since the feature has to be manually enabled, there is no risk during the introduction of the feature. 

- Potential issues are of high CPU usage, but we can improve on this as we see such cases. In the meantime, such users can disable the feature. 

## Acceptance Criteria

- Enable the option. Hover over a link and the page will start preloading. 

- Simply moving the mouse swiftly over the link shouldn’t start the preloading. 

- On a mobile device, preloading starts when the user touches the link before releasing it. Can be tested via console on browser stack.   

**Testing notes:** 

- Cross browser testing is recommended for this to make sure we don’t break things on different browsers. 

## Documentation and Translations

- Create new documentation to describe this feature and it’s benefits.

### PR DESCRIPTION
**Pull request by Tabrisrp**
Closes epic #2898 

Add a new option to the settings page to enable "Preload Links". When this option is enabled, an inline script is added to the footer of the page. This script detects when a user is going to click on a link to a page, and sends a prefetch request for the corresponding URL, to make the perceived loading of it as instant as possible.

### Acceptance Criteria

- Enable the option. Hover over a link and the page will start preloading. 
- Simply moving the mouse swiftly over the link shouldn’t start the preloading. 
- On a mobile device, preloading starts when the user touches the link before releasing it. Can be tested via console on browser stack.

When the preload link option is turned on, and the following plugins are also enabled, a notice is displayed in the backend:

- [Flying Pages](https://wordpress.org/plugins/flying-pages/)
- [Instant.page](https://wordpress.org/plugins/instant-page/)
- [Quicklink](https://wordpress.org/plugins/quicklink/)

**Testing notes:** 

- Cross browser testing is recommended for this to make sure we don’t break things on different browsers.

### CODE REVIEW

**Comment by Tabrisrp** [comment]
@hellofromtonya can you have a look at what David reported?

**Comment by hellofromtonya** [comment]
>The preload link script process links for images or any file extension

The other plugins from which this was modeled are doing a preload for the images and any file extension. This isn't right though. It shouldn't. I'll fix that part as lazyload should be handling the images and not this script.

**Comment by hellofromtonya** [comment]
>5. Preload links doesn't work on Safari.

"prefetch" isn't supported in Safari. [See caniuse.com for more info](https://caniuse.com/#feat=link-rel-prefetch).

**Comment by hellofromtonya** [comment]
Fixed:

- [x] Exclude external links
- [x] Exclude images

Maybe fixed:
- [x] IE11 and Edge. Opened up the babel to latest 2 browser versions

Can't fix:
- Does not work on Safari. 👈 prefetch is not supported on Safari, iOS Safari, or Opera mini

**Comment by Davidacu** [comment]
Exclude external links - 👍 
 Exclude images - 👍 
 Support for IE and Edge - Still not working. @hellofromtonya

**Comment by GeekPress** [comment]
> Support for IE and Edge - Still not working. 

@hellofromtonya I don't know if it will help the Edge compatibility, but we can forget the IE compatibility. It represents only 1.28% I don't think we need to have a more complex JS code for that.

**Comment by hellofromtonya** [comment]
@GeekPress Yes, that helps. We have IE/Edge issues elsewhere in Rocket and LazyLoad too. We can take that as a general topic to fix globally.

**Comment by GeekPress** [comment]
@hellofromtonya Does IE & Edge compatibilities require different piece of code for each one?

**Comment by hellofromtonya** [comment]
>Does IE & Edge compatibilities require different piece of code for each one?

Yes, Microsoft in general takes a different approach and slight changes in the events and/or handling. In theory, Babel (the tool we use to transpile the JS) should handle it. I suspect though something else is going on, something that doesn't translate over when running that tool. 

I'd have to debug this in a Windows container/simulator, i.e. which I have. I'll see if I can resolve it quickly. If no, we'll holistically look at our scripts within a Windows environment to ensure compatiblity.

**Comment by hellofromtonya** [comment]
>Support for IE and Edge - Still not working.

@Davidacu For now, we'll provide IE or Edge support.

**Comment by hellofromtonya** [comment]
@Davidacu @Tabrisrp There's one more thing this feature must do:

1. Exclude URLs with query string. 

I'm working on this now.

**Comment by hellofromtonya** [comment]
@Davidacu Can you test this PR with a shopping cart in WooCommerce and EDD? We need to know the following scenario:

1. Let’s imagine I’m visiting an e-commerce website.
2. I don’t have anything into my cart.
3. For an unknown reason (:D) I’m hovering the cart button, but don't click it.
4. The cart page is preloaded via this feature.
5. After that I decide to add an item to cart because I’m on a product category page where I can add quickly in one-click an item.
6. Then after adding the item into the cart, I have a pop-in to let me decide: continue or go to cart.
7. Then I go to the cart.

What happens? Does the new item appear in the cart? Are the items correct in the cart?

**Comment by hellofromtonya** [comment]
@Davidacu @Tabrisrp Updated the script to exclude preloading/prefetching links with a query string `?` such as:

```html
<a href="https://rocketlabsqa.dev/preload-links-query-string?product=shirt&color=blue&newuser&size=me">Query string with product=shirt&color=blue&newuser&size=m</a></li>
	<li><a href="https://rocketlabsqa.dev/preload-links-query-string?hello">Query string ?hello</a></li>
```

@Davidacu Can you add this one to your test cases please?

**Comment by hellofromtonya** [comment]
@Davidacu @Tabrisrp @GeekPress Ran tests on the Edge browser. 

1. The check for if the browser supports `prefetch` fails.
2. Disabling the browser check in our script. It does add the link prefetch element into the HTML. But the browser does not go and fetch it.
3. This time, added a link prefetch element into the `<head>` as part of a template, i.e. to see if it'd prefetch when the page loads. Nope. That didn't work either.

Looks like the Edge browser does not support link prefetch. Bummer. I did not test it on IE.

For now, we'll need to document that IE, Edge, iOS Safari, Safari, and Opera Mini are not compatible with this feature. We'll explore more in the future.

**Comment by Davidacu** [comment]
In general, I'm seeing unexpected behavior from WP Rocket when adding products to the cart in Woocomerce.
I added the cases with the query strings, but the items are still being prefetched @hellofromtonya .

**Issue 1** (Not related to preload-links)
1. Go to https://rocketlabsqa.dev/wp-woocomerce
2. Click on "Cart" -> Cart is empty
3. Back to home, add a product
4. Click on "Cart" 

**Expected**:
1 product on the cart
**Actual**:
Cart is empty
https://screencast-o-matic.com/watch/cYj2qQGCXX

**Issue 2** (preload-links related)
1. Go to https://rocketlabsqa.dev/wp-woocomerce
2. Click on "Cart" -> Cart is empty
3. Hover over "Add to cart" button for a few products
4. Finally, add one product by click on "Add to cart" button

**Expected**:
1 product on the cart
**Actual**:
All products that the user hovered over are in the cart.
https://screencast-o-matic.com/watch/cYj2qIGClA

**Comment by Davidacu** [comment]
Please disregard issue #1 @hellofromtonya @Tabrisrp , it was a configuration issue in Woocomerce settings.
issue #2 is still actual

**Comment by hellofromtonya** [comment]
>I added the cases with the query strings, but the items are still being prefetched

[Just checked the HTML here](https://rocketlabsqa.dev/wp-woocomerce/) and it does have the latest script in it. Did you grab the latest changes?

**Comment by hellofromtonya** [comment]
>I added the cases with the query strings, but the items are still being prefetched

[Just checked the HTML here](https://rocketlabsqa.dev/wp-woocomerce/) and it doesn't have the latest script in it. @Davidacu Did you grab the latest changes?

**Comment by hellofromtonya** [comment]
@Davidacu Issue 2 above is also related to the query string. [The newly revised script in this PR](https://github.com/wp-media/wp-rocket/pull/2989/commits/218d9230a071daa976a823343dbebffbc0a02fd4) hasn't been loaded onto the QA site yet. All of the `Add to cart` buttons have query string `?add-to-cart=id`. Once you get the updated script onto the QA site, please retest this one. Thanks.

**Comment by Davidacu** [comment]
>Did you grab the latest changes? 
Looks like I installed the wrong build 🤦‍♂️ 
re-testing now... (:

**Comment by hellofromtonya** [comment]
@Davidacu Sorry about that. Doh. I forgot to push the new compiled `.min.js` script. Doh. It's here now.

**Comment by Davidacu** [comment]
- I can confirm that now we don't prefetch links with query strings 👍 
- In Woocomerce when hovering over the cart link page, and after that adding a product and going to the cart, retrieves the correct items and quantity. 👍 
- In EDD though there seems to be an issue, when hovering over the cart, it prefetches the checkout page, if you add a product to the cart, and then go to checkout, it serves the prefetched cache checkout page, showing an empty cart.

To Reproduce:
1. Go to https://rocketlabsqa.dev/downloads/imagify-plugin/
2. Hover over the cart icon
3. Click on "Purchase" button
4. Click on "Checkout" button

**Expected:**
It should show the imagify plugin download
**Actual:**
It shows the empty cart.

**Comment by hellofromtonya** [comment]
Issue in EDD @Davidacu found is due to the `<a href="#"...>` in the EDD "Purchase" button. This anchor is used by EDD as a placeholder for an ajax call. It should be excluded as it's the same page.

https://github.com/wp-media/wp-rocket/pull/2989/commits/d164747b486b11630f6a6b99ba80ef21a52ddf91 does the following:

- [x] Excludes any URLs that have an anchor `#`.
- [x] Checks the protocol and excludes any URLs that are: `javascript:`, `mailto:` or `data:`. 

Test cases I used:

```html
<li><a href="#">#</a></li>
<li><a href="javascript:void(0)">javascript:void(0)</a></li>
<li><a href="mailto:bob@gmail.com">email</a></li>
<li><a href="data:text/plain;base64,SGVsbG8sIFdvcmxkIQ==">data</a></li>
```

**Comment by hellofromtonya** [comment]
https://github.com/wp-media/wp-rocket/pull/2989/commits/1b9f48ba0ac68a5216b038177011c9c63b1f01eb excludes links with `"checkout"` or `"cart"`.

**Comment by hellofromtonya** [comment]
New changes: Feature now excludes the following:

- [x] Pages:  add-to-cart`, `cart`, `checkout`, `wp-admin`, `wp-login.php`

**Comment by hellofromtonya** [comment]
@Davidacu This PR is ready to re-do QA.

**Comment by Davidacu** [comment]
- Issue in EDD with the cart has been fixed by excluding the `/checkout` from being prefetched. 👍 
- I can confirm the other excluded patterns are not prefetched.

All looking good! 🚀 
Find the latest test execution results attached. 100% Pass rate.
[testrail-report-17.pdf](https://github.com/wp-media/wp-rocket/files/5077865/testrail-report-17.pdf)

**Comment by GeekPress** [comment]
@hellofromtonya About the cart and checkout exclusion, we can’t trust the default slug.

A user can edit the default slug. And in case of multi-lingual websites, there are a different slug for each languages.

I guess our only trustable solution is to use the list of auto excluded pages from the cache.

**Comment by hellofromtonya** [comment]
@GeekPress Yes, late Friday, @Tabrisrp and I were talking about next steps, i.e. passing "exclude" URLs/page slugs to the script.

File: inc/Engine/Preload/Links/Subscriber.php [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+		wp_localize_script(
+			'rocket-preload-links',
+			'RocketPreloadLinksConfig',
+			[
+				'excludeUris'       => $this->get_uris_to_exclude( $use_trailing_slash ),
+				'usesTrailingSlash' => $use_trailing_slash,
+				'imageExtensions'   => 'jpg|jpeg|gif|png|tiff|bmp|webp|avif',
+			]
+		);
+	}
```
Comment (hellofromtonya) [inline inc/Engine/Preload/Links/Subscriber.php:107]: @Tabrisrp and @engahmeds3ed Please review the changes made here. What's going on?

1. Passing a configuration object to the `preload-links` script which includes:
     - The URLs to exclude.
     - Whether the permalink structure has a trailing slash or not (this is used in the script to properly format link's href for processing
     - Image extensions 
             - Why? To reduce the size of the script and give us the ability to change the configuration without touching the script. 
2. In the `get_uris_to_exclude()` method, we're now adding the admin, login, and logout URIs too.

**Comment by hellofromtonya** [comment]
@Tabrisrp and @engahmeds3ed Please review the changes and run it for yourself to ensure everything works properly. If yes, ping @Davidacu to re-do the QA.

**Comment by hellofromtonya** [comment]
@Davidacu What's changed since the last QA?

- Moved the hard-coded configurations (like image extensions and excluded page URLs) to the server. 
This configuration is then passed to the script. A fair bit of the script was rewritten to handle this change.

- We were getting HTTP 301 when a hover or clicked link did not have the trailing slash when the permalink structure required it. Fixed ✅  
For example: `<a href="/hello-world">` results in a 301 and then retry for `/hello-world/`. This change resulted in more handling and new script code. Need to validate it.

**Comment by hellofromtonya** [comment]
Update:

1. Folder subdomain sites: 
      - There was a grouping problem in the exclude URIs regex. @Tabrisrp resolved it.
      - `window.location.origin` did not work for a folder subdomain. Fixed by passing the site's URL as part of the configuration
2. The exclude URI regex pattern for checkout, cart, and my-account included `/(.*)` at the end of each URI. With regex in JS, that didn't work. Changed to strip that pattern off of the end of URIs.

## QA Testing

- [x] [Works with WooCommerce](https://rocketlabsqa.dev/wp-woocomerce/shop/). Confirmed the `/cart`, `/checkout`, and `/my-account` links are excluded and not preloaded. ✅ 
- [x] [Works with EDD](https://rocketlabsqa.dev/).  Confirmed `checkout` is excluded and not preloaded. ✅

**Pull request merged**